src/nvim/eval.c

```diff
@@ -21757,3 +21971,81 @@ bool eval_has_provider(char *name)
 
   return false;
 }
+
+static DictWatcher *dictwatcher_node_data(QUEUE *q)
+{
+  return QUEUE_DATA(q, DictWatcher, node);
+}
+
+static void dictwatcher_notify(dict_T *dict, const char *key,
+    typval_T *newtv, typval_T *oldtv)
+{
+  typval_T argv[2];
+  for (size_t i = 0; i < ARRAY_SIZE(argv); i++) {
+    init_tv(argv + i);
+  }
+
+  argv[0].v_type = VAR_STRING;
+  argv[0].vval.v_string = (uint8_t *)xstrdup(key);
+  argv[1].v_type = VAR_DICT;
+  argv[1].vval.v_dict = dict_alloc();
+  argv[1].vval.v_dict->dv_refcount++;
+
+  if (newtv) {
+    dictitem_T *v = dictitem_alloc((uint8_t *)"new");
+    copy_tv(newtv, &v->di_tv);
+    dict_add(argv[1].vval.v_dict, v);
+  }
+
+  if (oldtv) {
+    dictitem_T *v = dictitem_alloc((uint8_t *)"old");
+    copy_tv(oldtv, &v->di_tv);
+    dict_add(argv[1].vval.v_dict, v);
+  }
+
+  typval_T rettv;
+
+  QUEUE *w;
+  QUEUE_FOREACH(w, &dict->watchers) {
```

Comment:
>  Hash servers much better here then a queue with O(n) querying. You can’t have any patterns in this case, but plugins usually know all of the possible settings and for others there are two things:

It is theoretically possible to build a DFA for matching all patterns installed in a dictionary in a single walk over the key. In this case the DFA would have to be updated after every `watcher{add,del}`), but the number of watchers would not affect match performance.